### PR TITLE
[Components with more than 1 service] LPS-187067 Refactor components from frontend-js and frontend-theme modules

### PR DESCRIPTION
**Pull request by MrJohn1911**
Ticket: https://issues.liferay.com/browse/LPS-187067

This pull does not contains the component JSBundleConfigServlet because it requires more of work.

### CODE REVIEW

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

File: modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/util/JSImportMapsExtenderTopHeadDynamicIncludeUtil.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @author Joao Victor Alves
+ */
+public class JSImportMapsExtenderTopHeadDynamicIncludeUtil {
+
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/util/JSImportMapsExtenderTopHeadDynamicIncludeUtil.java:32]: Hey @MrJohn1911, please rename to `JSImportMapsRegistryUtil`. Thx

File: modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/modules/extender/npm/LastModifiedRegisterUtil.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * @author Joao Victor Alves
+ */
+public class LastModifiedRegisterUtil {
+
+	public static void accumulateAndGetLastModified(long lastModified) {
+		_lastModified.accumulateAndGet(lastModified, Math::max);
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/modules/extender/npm/LastModifiedRegisterUtil.java:25]: @MrJohn1911, rename to `updateLastModifed` instead.

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesNPMRegistryUpdatesListener.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@Component(
+	configurationPid = "com.liferay.frontend.js.loader.modules.extender.internal.configuration.Details",
+	service = NPMRegistryUpdatesListener.class
+)
+public class JSResolveModulesNPMRegistryUpdatesListener
+	implements NPMRegistryUpdatesListener {
+
+	@Override
+	public void onAfterUpdate() {
+		String hash = String.valueOf(UUID.randomUUID());
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesNPMRegistryUpdatesListener.java:37]: Hey @MrJohn1911, please move this logic to a method into the `JSResolveModulesRegistryUtil` and call this new method here and in the `activate` one from the `JSResolveModulesServlet`. Thx

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/util/JSResolveModulesRegistryUtil.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	public static String getExpectedPathInfo() {
+		return _expectedPathInfo;
+	}
+
+	public static String getUrl() {
+		return _url;
+	}
+
+	public static void setExpectedPathInfo(String expectedPathInfo) {
+		_expectedPathInfo = expectedPathInfo;
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/util/JSResolveModulesRegistryUtil.java:31]: please remove this set method

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/util/JSResolveModulesRegistryUtil.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	public static void setUrl(String url) {
+		_url = url;
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/util/JSResolveModulesRegistryUtil.java:35]: please remove this set method

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/util/JSResolveModulesRegistryUtil.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * details.
+ */
+
+package com.liferay.frontend.js.loader.modules.extender.internal.servlet.util;
+
+/**
+ * @author Joao Victor Alves
+ */
+public class JSResolveModulesRegistryUtil {
+
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/util/JSResolveModulesRegistryUtil.java:21]: rename to `JSLoaderModulesUtil` instead

**Review by rafaprax — CHANGES_REQUESTED** [review]

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:stop

**Comment by MrJohn1911** [comment]
ci:test:sf

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

File: modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/modules/extender/npm/LastModifiedRegisterUtil.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+package com.liferay.frontend.js.web.internal.modules.extender.npm;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * @author Joao Victor Alves
+ */
+public class LastModifiedRegisterUtil {
+
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/modules/extender/npm/LastModifiedRegisterUtil.java:23]: @MrJohn1911  rename class to `LastModifiedUtil`

File: modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/modules/extender/npm/LastModifiedRegisterUtil.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+		return _lastModified.get();
+	}
+
+	public static void setLastModified(long lastModified) {
+		_lastModified.set(lastModified);
+	}
+
+	public static void updateLastModified(long lastModified) {
+		_lastModified.accumulateAndGet(lastModified, Math::max);
+	}
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/modules/extender/npm/LastModifiedRegisterUtil.java:34]: @MrJohn1911 rename to `accumulateAndGetLastModified`, please.

**Review by rafaprax — CHANGES_REQUESTED** [review]

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by zhangyan0701** [comment]
Issue found:

**In Client Extension portlet, there will be a duplicate element created in page HTML source for the drop down menu items after clicking back icon.**

Reproduce steps:
1. Go to Global Menu > Applications > Custom App > Client Extensions.
2. Click Plus icon to open the drop down menu.
3. Open browser developer tool by clicking F12.
4. Search for element of the Add IFrame item in HTML: `//button[@class='dropdown-item' and contains(text(),'IFrame')]`.
5. Assert that there is only one result.
6. Click another item, such as Custom Element.
7. Do nothing then click back icon.
8. Open the drop down menu again, search for the element of the Add IFrame item.
9. Assert there is one more result in HTML. One is not display on page.
10. Repeat steps 6-8, assert that every time after clicking the back icon, there is one more drop down menu item element in HTML.

https://github.com/liferay-core-infra/liferay-portal/assets/25576873/893ab9cf-fdbd-477e-87f1-01783855d727

**Comment by MrJohn1911** [comment]
ci:test:sf

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by zhangyan0701** [comment]
Hi @MrJohn1911 

The issue in this comment https://github.com/liferay-core-infra/liferay-portal/pull/4744#issuecomment-1596820933 still happens.

Affected test:
[LocalFile.ClientExtension#IframeTypeCanDisplayProperFields](https://testray.liferay.com/home/-/testray/case_results/2355897567)

**Pull request closed without merging**